R/pipe_consistency_linter.R

```diff
@@ -0,0 +1,79 @@
+#' Pipe consistency linter
+#'
+#' Check that pipe operators are consistently used throughout a file. Optionally
+#' specify which pipe operator to use.
+#'
+#' @param pipe Which pipe operator to use (either `"%>%"` or `"|>"`). By default
+#' (`"auto"``), the linter has no preference but will check that the same pipe
+#' operator is used throughout the file.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "1:3 |> mean() %>% as.character()",
+#'   linters = pipe_consistency_linter()
+#' )
+#'
+#' lint(
+#'   text = "1:3 %>% mean() %>% as.character()",
+#'   linters = pipe_consistency_linter("|>")
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "1:3 %>% mean() %>% as.character()",
+#'   linters = pipe_consistency_linter()
+#' )
+#'
+#' lint(
+#'  text = "1:3 |> mean() |> as.character()",
+#'  linters = pipe_consistency_linter()
+#' )
+#' @evalRd rd_tags("pipe_consistency_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+pipe_consistency_linter <- function(pipe = c("auto", "%>%", "|>")) {
+  pipe <- match.arg(pipe)
+
+  xpath_magrittr <- "//SPECIAL[text() = '%>%']"
+  xpath_native <- "//PIPE"
+
+  Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, "file")) {
+      return(list())
+    }
+
+    xml <- source_expression$full_xml_parsed_content
+
+    match_magrittr <- xml2::xml_find_all(xml, xpath_magrittr)
```

Comment:
nit: for efficiency, we might prefer:

```r
match_magrittr <- NULL
match_native <- NULL
if (pipe %in% c("auto", "|>")) {
  match_magrittr <- xml2::xml_find_all(xml, xpath_magrittr)
}
if (pipe %in% c("auto", "%>%")) {
  match_native <- xml2::xml_find_all(xml, xpath_native)
}
```

to avoid running the XPath if it's not needed for the current `pipe`.

Not strictly necessary since this should be a relatively fast XPath & there's a slight reduction in readability, WDYT?

Reply:
Very rough benchmark over a 1000 line file I generated with 3000 pipe operators.

```
  expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result memory              time       gc
  <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list> <list>              <list>     <list>  
1 magrittr      719µs  786.2µs     1170.   14.12KB     0      585     0      500ms <NULL> <Rprofmem [8 × 3]>  <bench_tm> <tibble>
2 native        223µs  253.6µs     3467.    5.69KB     2.02  1713     1      494ms <NULL> <Rprofmem [8 × 3]>  <bench_tm> <tibble>
3 both          950µs   1.01ms      899.   19.81KB     2.03   443     1      493ms <NULL> <Rprofmem [16 × 3]> <bench_tm> <tibble>
```

The magrittr matching is clearly slower, but still pretty fast. Is it enough to justify the hit to readability?

Reply:
I'm somewhat indifferent, I'll leave it up to you :)

Reply:
I think it's tidier as it is, so going to leave it 🙂